src/Apps/Order/Dialogs.tsx

```diff
@@ -0,0 +1,157 @@
+  // We are quite unlikely to need this, but let's err on the side of caution
+  private async maybeForceCloseExistingDialog() {
+    if (this.state.props.show) {
+      await this.state.forceCloseDialog()
+    }
+  }
+
+  showAcceptDialog = async ({
+    title,
+    message,
+    continueButtonText = "Continue",
+    cancelButtonText = "Cancel",
+  }: {
+    title: React.ReactNode
+    message: React.ReactNode
+    continueButtonText?: string
+    cancelButtonText?: string
+  }): Promise<{ accepted: boolean }> => {
+    await this.maybeForceCloseExistingDialog()
+
+    return new Promise<{ accepted: boolean }>(resolve => {
+      const accept = async () => {
+        await this.closeDialog()
+        resolve({ accepted: true })
+      }
+      const reject = async () => {
+        await this.closeDialog()
+        resolve({ accepted: false })
+      }
+
+      this.setState({
+        props: {
+          show: true,
+          heading: title,
+          detail: message,
+          primaryCta: {
+            text: continueButtonText,
+            action: accept,
+          },
+          secondaryCta: {
+            text: cancelButtonText,
+            action: reject,
+          },
+          onClose: reject,
+        },
+        forceCloseDialog: reject,
+      })
+    })
+  }
+
+  showErrorDialog = async ({
+    title = "An error occurred",
+    supportEmail = "orders@artsy.net",
+    message = (
+      <>
+        Something went wrong. Please try again or contact{" "}
+        <Link href={`mailto:${supportEmail}}`}>{supportEmail}</Link>.
+      </>
+    ),
+    continueButtonText = "Continue",
+  }: {
+    title?: string
+    message?: React.ReactNode
+    supportEmail?: string
+    continueButtonText?: string
+  }): Promise<void> => {
+    await this.maybeForceCloseExistingDialog()
+
+    return new Promise<void>(resolve => {
+      const onContinue = async () => {
+        await this.closeDialog()
+        resolve()
+      }
+
+      this.setState({
+        props: {
+          show: true,
+          heading: title,
+          detail: message,
+          primaryCta: {
+            text: continueButtonText,
+            action: onContinue,
+          },
+          onClose: onContinue,
+        },
+        forceCloseDialog: onContinue,
+      })
+    })
+  }
+}
+
+const extractDialogHelpers = ({
+  showAcceptDialog,
+  showErrorDialog,
+}: DialogContainer) => ({ showAcceptDialog, showErrorDialog })
+
+export type DialogHelpers = ReturnType<typeof extractDialogHelpers>
+
+// make it safe to inject dialogs on components nested within a page by
+// only injecting the actual modal component if it hasn't been injected already
+// otherwise you might end up with two or more instances of the modal appearing.
+const DialogContext = React.createContext({ isModalInTree: false })
+
+/**
+ * Injects the `dialogs` prop into the given page component
+ * @param Component
+ */
+export function injectDialogs<R extends { dialogs: DialogHelpers }>(
```

Comment:
While I think this is a nice abstraction in terms of safety, I wonder if we might want to approach this from two (possibly) simpler angles: 
1) Move `<Subscribe to={[DialogContainer]}>` and `ModalDialog` into the [root `OrderApp`](https://github.com/artsy/reaction/blob/master/src/Apps/Order/OrderApp.tsx#L91) render method. Then, from some area of the subtree, we can show / hide by passing the dialog definition to unstated which will update and then render the modal directly from `OrderApp`. Then we never have to check for dups, or maintain the overhead of additional contexts. 
2) Pass a state change to unstated which then attaches a modal via a [portal](https://reactjs.org/docs/portals.html)

The above examples are just suggestions and I may be missing something. In any case the below is clear.

Reply:
Or better, rather than `showAcceptDialog` and `showErrorDialog` we can keep things generic and have the call-sites manage this info, instead passing down `showDialog`. If `error` and `accept` are used in numerous places then those values can be shared across files via a `constants` file (or something).

Reply:
I had done 1 originally but couldn't get access to the modals in enzyme without wrapping the routes in an `<OrderApp>` which seemed to cause some unsavoury issues that I thought seemed more complex to tackle than doing it like this.

Now that you bring it up I do see a flaw in this pattern: if you have two siblings wrapped with injectDialogs but not the parent, then you still end up with duplicate ModalDialogs in the dom. Unlikely to happen and easy to spot during development, but still not ideal.

No idea about using a portal. I suspect that might cause enzyme troubles too?

I could try going back to something like 1 but abstracting the Subscribe+ModalDialog bit into a <ProvideDialogs> component that can be used easily in both OrderApp.tsx and test files. What do you think?

Reply:
> Or better, rather than showAcceptDialog and showErrorDialog being managed at the state level, we can keep things generic and have the call-sites manage this info, instead passing down showDialog. If error and accept are used in numerous places then those values can be shared across files via a constants file (or something) from within the order app.

Oops sorry i missed this message earlier.

Yeah both are used in a bunch of places, we defo don't wanna be specifying the whole `ModalDialogProps` every time we show a dialog. So moving that formatting/interaction logic for the two types of dialog somewhere else would be good for tree shaking if the dialogs are ever used in other apps, but would also be a more cumbersome API as far as I can tell. Not worth it just yet IMO.

Reply:
I would encourage keeping it simple and finding a way to stick with one main modal at the root and then use unstated to render with updated props / hide but not a blocker if you feel strongly about current approach 👍

Reply:
Managed to do what you had in mind here.

- Got rid of the context business and figured out a way to insert the actual `ModalDialog` instance into OrderApp and also use it in tests. 
- Now the injected dialog helpers are just `show` and `hide` and all the logic about accept vs error is separate.